### PR TITLE
feat(eslint-plugin): [no-shadow] add shadowed variable location to the error message

### PR DESCRIPTION
**Pull request by bradzacher**
## PR Checklist

-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

I noticed that we had missed https://github.com/eslint/eslint/pull/13841. This is a pretty valuable upstream change which greatly improves the error message.

Whilst I was here I also followed the same test structure for `no-unused-vars` where we separate the tests from upstream eslint into a separate file from our local tests.
I did this now because I noticed that someone had added tests for our logic to the "eslint tests" section - so when I pasted the upstream tests I deleted a bunch of tests!
It shouldn't be an issue in future.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-shadow.ts [hunk 1]
```diff
@@ -595,12 +619,26 @@ export default util.createRule<Options, MessageIds>({
...
+                },
+              } as const);
+
           context.report({
             node: variable.identifiers[0],
-            messageId: 'noShadow',
-            data: {
-              name: variable.name,
-            },
+            ...report,
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-shadow.ts:641]: Nit: the ternary could be inlined in here. Then there would be no need for the `as const`s.

```ts
context.report({
    node: variable.identifiers[0],
        ...(location.global
            ? {
                messageId: 'noShadowGlobal',
            // ...
```

File: packages/eslint-plugin/src/rules/no-shadow.ts [hunk 2]
```diff
@@ -517,6 +519,28 @@ export default util.createRule<Options, MessageIds>({
...
+        return {
+          global: false,
+          line: identifier.loc.start.line,
+          column: identifier.loc.start.column + 1,
+        };
+      } else {
+        return {
+          global: true,
+        };
+      }
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-shadow.ts:541]: Nitty nit: could be simplified to `return identifier ? { ... } : { ... };`.

**Review by JoshuaKGoldberg — APPROVED** [review]
Super, thank you!

File: packages/eslint-plugin/src/rules/no-shadow.ts (hunk 2 above)
Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-shadow.ts:541]: you really do love your ternaries 😄 
*personally* I prefer to only use ternaries for when I want an expression.
I find in this case where I can do everything with statements - using an if statement is clearer and cleaner.

**Pull request merged**